### PR TITLE
WPB-663 servantify brig provider service API

### PR DESCRIPTION
**Pull request by battermann**
https://wearezeta.atlassian.net/browse/WPB-663

Do not merge before the mandarin release as this could potentially break QA's automation.

There has been a problem reported by QA with a previous attempt of this, and I now checked every endpoint again, making sure they are all covered by a test. However, I wasn't able to reproduce the issue that QA ran into:

```
https://staging-nginz-https.zinfra.io/teams/e4898f4d-160a-489c-9484-b3cd830bd84f/services/whitelist (404): {"code":404,"label":"no-endpoint","message":"The requested endpoint does not exist"} (404)
```

Even an integration test that call this via nginz passes. It's not the content type or accept headers, either. Also, it works both with versioned and unverified paths.

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: libs/wire-api/src/Wire/API/Provider/Service.hs [hunk 1]
```diff
@@ -300,6 +333,7 @@ data ServiceProfilePage = ServiceProfilePage
   }
   deriving stock (Eq, Show, Generic)
   deriving (Arbitrary) via (GenericUniform ServiceProfilePage)
+  deriving (S.ToSchema) via (Schema ServiceProfilePage)
 
 instance ToJSON ServiceProfilePage where
```
Comment (elland) [inline libs/wire-api/src/Wire/API/Provider/Service.hs:338]: Aren't these instances redundant now? We could just have them in terms of Schema, right?

File: services/brig/test/integration/API/Provider.hs [hunk 2]
```diff
@@ -97,11 +97,12 @@ import Wire.API.Team.Feature (featureNameBS)
 import Wire.API.Team.Feature qualified as Public
 import Wire.API.Team.Permission
 import Wire.API.User as User hiding (EmailUpdate, PasswordChange, mkName)
+import Wire.API.User.Auth (CookieType (..))
```
Comment (elland) [inline services/brig/test/integration/API/Provider.hs:100]: Not sure why this change.

Comment (battermann) [reply services/brig/test/integration/API/Provider.hs:100]: I added a test which uses the nginz endpoint instead of brig (trying to reproduce the failing test that QA reported). I thought it doesn't hurt to leave this test in here.

File: libs/wire-api/src/Wire/API/Provider/Service.hs (hunk 1 above)
Comment (battermann) [reply libs/wire-api/src/Wire/API/Provider/Service.hs:338]: I am bot aware.

File: libs/wire-api/src/Wire/API/Provider.hs [hunk 3]
```diff
@@ -56,7 +56,7 @@ import Data.Id
 import Data.Misc (HttpsUrl (..), PlainTextPassword6, PlainTextPassword8)
 import Data.Range
 import Data.Schema
-import Data.Swagger qualified as S
+import Data.OpenApi qualified as S
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/Provider.hs]: was that an oversight?  shouldn't Data.Swagger be unavailable at this point?  (probably a topic for a separate PR.)

File: libs/wire-api/src/Wire/API/Provider/Service.hs [hunk 4]
```diff
@@ -205,6 +208,22 @@ data Service = Service
   }
   deriving stock (Eq, Show, Generic)
   deriving (Arbitrary) via (GenericUniform Service)
+  deriving (S.ToSchema) via (Schema Service)
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/Provider/Service.hs:211]: please also remove the manual `{To,From}JSON` instances, or leave a comment why that's hard.

File: libs/wire-api/src/Wire/API/Provider/Service.hs [hunk 5]
```diff
@@ -265,6 +284,7 @@ data ServiceProfile = ServiceProfile
   }
   deriving stock (Eq, Show, Generic)
   deriving (Arbitrary) via (GenericUniform ServiceProfile)
+  deriving (S.ToSchema) via (Schema ServiceProfile)
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/Provider/Service.hs:287]: ToJSON, FromJSON

File: libs/wire-api/src/Wire/API/Provider/Service.hs [hunk 6]
```diff
@@ -330,6 +358,20 @@ data NewService = NewService
   }
   deriving stock (Eq, Show, Generic)
   deriving (Arbitrary) via (GenericUniform NewService)
+  deriving (S.ToSchema) via (Schema NewService)
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/Provider/Service.hs:361]: ToJSON, FromJSON

File: libs/wire-api/src/Wire/API/Provider/Service.hs [hunk 7]
```diff
@@ -366,6 +408,14 @@ data NewServiceResponse = NewServiceResponse
   }
   deriving stock (Eq, Show, Generic)
   deriving (Arbitrary) via (GenericUniform NewServiceResponse)
+  deriving (S.ToSchema) via (Schema NewServiceResponse)
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/Provider/Service.hs:411]: ToJSON, FromJSON

File: libs/wire-api/src/Wire/API/Provider/Service.hs [hunk 8]
```diff
@@ -393,6 +443,17 @@ data UpdateService = UpdateService
   }
   deriving stock (Eq, Show, Generic)
   deriving (Arbitrary) via (GenericUniform UpdateService)
+  deriving (S.ToSchema) via (Schema UpdateService)
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/Provider/Service.hs:446]: ToJSON, FromJSON

File: libs/wire-api/src/Wire/API/Provider/Service.hs [hunk 9]
```diff
@@ -427,6 +488,17 @@ data UpdateServiceConn = UpdateServiceConn
   }
   deriving stock (Eq, Show, Generic)
   deriving (Arbitrary) via (GenericUniform UpdateServiceConn)
+  deriving (S.ToSchema) via (Schema UpdateServiceConn)
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/Provider/Service.hs:491]: ToJSON, FromJSON

File: libs/wire-api/src/Wire/API/Provider/Service.hs [hunk 10]
```diff
@@ -458,6 +530,13 @@ newtype DeleteService = DeleteService
   {deleteServicePassword :: PlainTextPassword6}
   deriving stock (Eq, Show)
   deriving newtype (Arbitrary)
+  deriving (S.ToSchema) via (Schema DeleteService)
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/Provider/Service.hs:533]: ToJSON, FromJSON

File: libs/wire-api/src/Wire/API/Provider/Service/Tag.hs [hunk 11]
```diff
@@ -1,4 +1,5 @@
 {-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/Provider/Service/Tag.hs:2]: why?  can it be avoided?

File: libs/wire-api/src/Wire/API/Routes/Public/Brig.hs [hunk 12]
```diff
@@ -1492,6 +1496,10 @@ type TeamsAPI =
                     (Respond 200 "Number of team members" TeamSize)
            )
 
+-- Plaintext doesn't ship a renderer for (), so we have an orphan for it
+instance MimeRender PlainText () where
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/Routes/Public/Brig.hs:1500]: What about `NoContent`?

File: services/nginz/integration-test/conf/nginz/nginx.conf [hunk 13]
```diff
@@ -244,6 +244,11 @@ http {
       proxy_pass http://brig;
     }
 
+    location ~* /teams/([^/]+)/services {
```
Comment (fisx) [inline services/nginz/integration-test/conf/nginz/nginx.conf:247]: it's not nice that we have so many config files that should in fact all be the same :(

File: libs/wire-api/src/Wire/API/Provider.hs (hunk 3 above)
Comment (elland) [reply libs/wire-api/src/Wire/API/Provider.hs]: We've moved away from it to openapi3 a while back.

File: libs/wire-api/src/Wire/API/Provider/Service/Tag.hs [hunk 14]
```diff
@@ -64,6 +69,13 @@ import Wire.Arbitrary (Arbitrary (..), GenericUniform (..))
 newtype ServiceTagList = ServiceTagList [ServiceTag]
   deriving stock (Eq, Ord, Show)
   deriving newtype (FromJSON, ToJSON, Arbitrary)
+  deriving (S.ToSchema) via (Schema ServiceTagList)
+
+_ServiceTagList :: Prism' ServiceTagList [ServiceTag]
+_ServiceTagList = prism ServiceTagList (\(ServiceTagList l) -> pure l)
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/Provider/Service/Tag.hs:75]: why is this not a lens?

File: libs/wire-api/src/Wire/API/Provider/Service/Tag.hs [hunk 15]
```diff
@@ -100,6 +112,7 @@ data ServiceTag
   | WeatherTag
   deriving stock (Eq, Show, Ord, Enum, Bounded, Generic)
   deriving (Arbitrary) via (GenericUniform ServiceTag)
+  deriving (S.ToSchema) via (Schema ServiceTag)
```
Comment (fisx) [inline libs/wire-api/src/Wire/API/Provider/Service/Tag.hs:115]: another FromJSON, ToJSON :)

**Review by fisx — APPROVED** [review]
